util/src/main/scala/org/ensime/util/Debouncer.scala

```diff
@@ -0,0 +1,118 @@
+// Copyright: 2010 - 2017 https://github.com/ensime/ensime-server/graphs
+// License: http://www.gnu.org/licenses/gpl-3.0.en.html
+package org.ensime.util
+
+import akka.actor.Scheduler
+import java.time.Clock
+import java.util.concurrent.atomic.{ AtomicBoolean, AtomicInteger }
+import org.slf4j.LoggerFactory
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration._
+
+/**
+ * Generic debouncer implementation
```

Comment:
Yes, I understood the reasons... could you please write that up in the comments so that anybody reading the code can understand why we did it this way. It is unexpected that we wouldn't just write this up as a reusable Actor since all the plumbing of the server is written in akka. It is very bad akka style to use closures because it allows state to leak across the concurrency model, so it requires justification / explanation. I think performance is a good justification.